src/services/services.ts

```diff
@@ -6259,6 +6373,187 @@ namespace ts {
+                            maybeAdd(getReferenceEntryFromNode(parent.initializer));
+                        }
+                        else if (isFunctionLike(parent) && parent.type === containingTypeReference && parent.body && parent.body.kind === SyntaxKind.Block) {
+                            forEachReturnStatement(<Block>parent.body, (returnStatement) => {
+                                if (returnStatement.expression && isImplementationExpression(returnStatement.expression)) {
+                                    maybeAdd(getReferenceEntryFromNode(returnStatement.expression));
+                                }
+                            });
+                        }
+                        else if (isTypeAssertionExpression(parent) && isImplementationExpression(parent.expression)) {
+                            maybeAdd(getReferenceEntryFromNode(parent.expression));
+                        }
+                    }
+                }
+
+                // Type nodes can contain multiple references to the same type. For example:
+                //      let x: Foo & (Foo & Bar) = ...
+                // Because we are returning the implementation locations and not the identifier locations,
+                // duplicate entries would be returned here as each of the type references is part of
+                // the same implementation. For that reason, check before we add a new entry
+                function maybeAdd(a: ReferenceEntry) {
+                    if (!forEach(result, b => a.fileName === b.fileName && a.textSpan.start === b.textSpan.start && a.textSpan.length === b.textSpan.length)) {
+                        result.push(a);
+                    }
+                }
+            }
+
+            function getSymbolsForComponentTypes(type: UnionOrIntersectionType, result: Symbol[] = []): Symbol[] {
+                for (const componentType of type.types) {
+                    if (componentType.symbol && componentType.symbol.getFlags() & (SymbolFlags.Class | SymbolFlags.Interface)) {
+                        result.push(componentType.symbol);
+                    }
+                    if (componentType.getFlags() & TypeFlags.UnionOrIntersection) {
+                        getSymbolsForComponentTypes(<UnionOrIntersectionType>componentType, result);
+                    }
+                }
+                return result;
+            }
+
+            function getContainingTypeReference(node: Node): Node {
+                let topLevelTypeReference: Node = undefined;
+
+                while (node) {
+                    if (isTypeNode(node)) {
+                        topLevelTypeReference = node;
+                    }
+                    node = node.parent;
+                }
+
+                return topLevelTypeReference;
+            }
+
+            function getContainingClassHeritageClause(node: Node): HeritageClause {

```

Comment:
The return value of this function is only ever used for this parent, so might as well just return the containing class?